src/Tasks/Unzip.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Threading;
+
+namespace Microsoft.Build.Tasks
+{
+    /// <summary>
+    /// Represents a task that can extract a .zip archive.
+    /// </summary>
+    public sealed class Unzip : TaskExtension, ICancelableTask
+    {
+        /// <summary>
+        /// Stores a <see cref="CancellationTokenSource"/> used for cancellation.
+        /// </summary>
+        private readonly CancellationTokenSource _cancellationToken = new CancellationTokenSource();
+
+        /// <summary>
+        /// Gets or sets a <see cref="ITaskItem"/> with a destination folder path to unzip the files to.
+        /// </summary>
+        [Required]
+        public ITaskItem DestinationFolder { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether read-only files should be overwritten.
+        /// </summary>
+        public bool OverwriteReadOnlyFiles { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether files should be skipped if the destination is unchanged.
+        /// </summary>
+        public bool SkipUnchangedFiles { get; set; } = true;
+
+        /// <summary>
+        /// Gets or sets an array of <see cref="ITaskItem"/> objects containing the paths to .zip archive files to unzip.
+        /// </summary>
+        [Required]
+        public ITaskItem[] SourceFiles { get; set; }
+
+        /// <inheritdoc cref="ICancelableTask.Cancel"/>
+        public void Cancel()
+        {
+            _cancellationToken.Cancel();
+        }
+
+        /// <inheritdoc cref="Task.Execute"/>
+        public override bool Execute()
+        {
+            DirectoryInfo destinationDirectory;
+            try
+            {
+                destinationDirectory = Directory.CreateDirectory(DestinationFolder.ItemSpec);
+            }
+            catch (Exception e)
+            {
+                Log.LogErrorFromResources("Unzip.ErrorCouldNotCreateDestinationDirectory", DestinationFolder.ItemSpec, e.Message);
+
+                return false;
+            }
+
+            BuildEngine3.Yield();
+
+            foreach (ITaskItem sourceFile in SourceFiles.TakeWhile(i => !_cancellationToken.IsCancellationRequested))
+            {
+                if (!File.Exists(sourceFile.ItemSpec))
+                {
+                    Log.LogErrorFromResources("Unzip.ErrorFileDoesNotExist", sourceFile.ItemSpec);
+                    continue;
```

Comment:
Why not eagerly exit to reduce the time it takes to finish the build on error?

Reply:
Ah, to show all errors at once rather than one at a time. The alternative is to complicate things and check everything first, and do the extraction second.

---
In reply to: [188788554](https://github.com/Microsoft/msbuild/pull/3291#discussion_r188788554) [](ancestors = 188788554)